### PR TITLE
Flask: Permission Gating

### PR DESCRIPTION
**Pull request by SamarthVP**
For #117

The endpoints are a pain to test, but one can set `LOGIN_DISABLED=True` in `config.py` and then pass `?user=x` to the endpoints to emulate users

Then mess with the db and see what comes out.

Testing PATCH `/api/datasets/:id`
```
curl -X PATCH \
    -H "Content-Type: application/json" \
    --data '{
            "input_hpf_path": "woohoo"
    }' \
    localhost:5000/api/datasets/3\?user=1
```

Testing PATCH `/api/participants/:id`
```
curl -X PATCH \
    -H "Content-Type: application/json" \
    --data '{
            "notes": "k"
    }' \
    localhost:5000/api/participants/3\?user=1
```

Testing PATCH `/api/families/:id`
```
curl -X PATCH \
    -H "Content-Type: application/json" \
    --data '{
            "family_codename": 5
    }' \
    localhost:5000/api/families/1\?user=1
```

Testing PATCH `/api/analyses/:id`
```
curl -X PATCH \
    -H "Content-Type: application/json" \
    --data '{
            "result_hpf_path": "nice one 2"
    }' \
    localhost:5000/api/analyses/2\?user=1
```

### CODE REVIEW

File: flask/app/participants.py [hunk 1]
```diff
@@ -15,19 +15,51 @@
...
 
     participants = [
         {
             **asdict(participant),
             "family_codename": participant.family.family_codename,
             "tissue_samples": [
-                {**asdict(tissue_sample), "datasets": tissue_sample.datasets}
+                {
+                    **asdict(tissue_sample),
+                    "datasets": (
```
Comment (SamarthVP) [inline flask/app/participants.py:45]: Multiple db queries works but runs the risk of being slow. I couldn't really think of a better way though...
Thoughts @kevinlul ?

File: flask/app/families.py [hunk 2]
```diff
@@ -11,14 +11,46 @@
 @login_required
 def families_list():
 
-    db_families = models.Family.query.options(
-        joinedload(models.Family.participants)
-    ).all()
+    db_families = (
+        models.Family.query.join(models.Participant)
```
Comment (kevinlul) [inline flask/app/families.py:15]: Before the first join, add `.options(contains_eager(models.Family.participants))` and make sure to import `contains_eager` from `sqlalchemy.orm` and see if this works for the rest.

File: flask/app/datasets.py [hunk 3]
```diff
@@ -17,6 +17,13 @@
 def datasets_list():
     db_datasets = (
         db.session.query(models.Dataset)
+        .join(models.groups_datasets_table)
+        .join(
+            models.users_groups_table,
+            models.groups_datasets_table.columns.group_id
+            == models.users_groups_table.columns.group_id,
+        )
+        .filter(models.users_groups_table.columns.user_id == current_user.user_id)
```
Comment (kevinlul) [inline flask/app/datasets.py:26]: We need to remember to ignore this if the current user is an admin

**Comment by SamarthVP** [comment]
Now that we've done the general `/api/*thing*` endpoints, I believe we need to extend this to the `/api/*thing*/:id` endpoints. Or is that not a concern because `/api/*thing*/:id` would only ever be called on `*things*` coming from the `/api/*thing*` endpoints?

**Comment by kevinlul** [comment]
It should be extended and the logic would be to 404 if the user doesn't have permission (maybe 403 if assuming a user identity as admin).

**Comment by SamarthVP** [comment]
I think 404 in all cases where permission is blocked makes the most sense

File: flask/app/participants.py [hunk 4]
```diff
@@ -76,15 +133,13 @@ def update_participant(id: int):
...
         return enum_error, 400
 
-    try:
-        table.updated_by = current_user.user_id
-    except:
-        pass  # LOGIN_DISABLED
+    if user_id:
+        participant.updated_by = user_id
 
     routes.transaction_or_abort(db.session.commit)
```
Comment (SamarthVP) [inline flask/app/participants.py:145]: Should return jsonify(participant) since table got renamed

File: flask/app/analyses.py [hunk 5]
```diff
@@ -24,65 +29,116 @@ def analyses_list():
...
-    return jsonify(analyses)
+    return jsonify(
+        [
+            {
+                **asdict(analysis),
+                "requester": requester and requester.username,
+                "updated_by": updated_by and updated_by.username,
+                "assignee": assignee and assignee.username,
+            }
+            for analysis, requester, updated_by, assignee in db_analyses
```
Comment (SamarthVP) [inline flask/app/analyses.py:72]:
```suggestion
            for analysis, requester, updated_by, assignee in analyses
```
Another one

**Comment by helrick** [comment]
lgtm! (based on currently defined permissions)

small item: non-admins are still able to see the admin page. not sure if that's in the scope of this issue though

**Review by helrick — APPROVED** [review]

**Pull request merged**